README.md
# NAME

MaxMind::DB::Common - Code shared by the MaxMind DB reader and writer modules

# VERSION

version 0.040000

# DESCRIPTION

This distribution provides some shared code for use by both the MaxMind DB
reader and writer Perl modules.

For now, the only piece documented for public consumption is
[MaxMind::DB::Metadata](https://metacpan.org/pod/MaxMind::DB::Metadata).

# VERSIONING POLICY

This module uses semantic versioning as described by
[http://semver.org/](http://semver.org/). Version numbers can be read as X.YYYZZZ, where X is the
major number, YYY is the minor number, and ZZZ is the patch number.

# SUPPORT

Please report all issues with this code using the GitHub issue tracker at
[https://github.com/maxmind/MaxMind-DB-Common-perl/issues](https://github.com/maxmind/MaxMind-DB-Common-perl/issues).

# AUTHORS

- Dave Rolsky <drolsky@maxmind.com>
- Olaf Alders <oalders@maxmind.com>

# CONTRIBUTOR

Greg Oschwald <goschwald@maxmind.com>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by MaxMind, Inc..

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
{{$NEXT}}

0.040001 2016-02-02

- Validation on the Int type used in the Metadata class has been loosened,
  which should fix MaxMind::DB::Reader test failures on some platforms.


0.040000 2015-04-29

- Use Math::BigInt rather than Math::Int128 as part of working towards making
  MaxMind::DB::Reader pure Perl.


0.031003 2014-03-10

- Added docs for MaxMind::DB::Metadata.

- Added two more exported constants, DATA_SECTION_SEPARATOR and
  METADATA_MARKER.


0.031002 2013-07-18

- Changes to make code work with Perl 5.8.x.


0.031001 2013-07-16

- Fixed a test failure with Perl 5.10.


0.031000 2013-07-16

- First release upon an unsuspecting world.


lib/Test/MaxMind/DB/Common/Data.pm
package Test::MaxMind::DB::Common::Data;

use strict;
use warnings;

our $VERSION = '0.040002';

use Exporter qw( import );
use Math::BigInt ();

our @EXPORT_OK = qw( test_cases_for );

sub test_cases_for {
    my $type = shift;

    my $method = '_' . $type;

    return __PACKAGE__->$method(@_);
}

## no critic (Subroutines::ProhibitUnusedPrivateSubroutines)
sub _array {
    my @one_value = (
        ['Foo'] => [
            0b00000001,
            0b00000100,
            (
                # Foo
                0b01000011,
                0b01000110, 0b01101111, 0b01101111
            )
        ],
    );

    my @two_values = (
        [
            'Foo',
            "\x{4eba}",
        ] => [
            0b00000010,
            0b00000100,
            (
                # Foo
                0b01000011,
                0b01000110, 0b01101111, 0b01101111
            ),
            (
                # \x{4eba}
                0b01000011,
                0b11100100, 0b10111010, 0b10111010
            ),
        ],
    );

    return [
        [] => [ 0b00000000, 0b00000100 ],
        @one_value,
        @two_values,
    ];
}

sub _bytes {
    my @bytes = (
        q{}   => [0b10000000],
        '1'   => [ 0b10000001, ord('1') ],
        '123' => [ 0b10000011, map { ord($_) } qw( 1 2 3 ) ],
        '123456789012345678901234567' => [
            0b10011011,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 )
        ],
        '1234567890123456789012345678' => [
            0b10011100,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 )
        ],
        '12345678901234567890123456789' => [
            0b10011101, 0b00000000,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 )
        ],
        '123456789012345678901234567890' => [
            0b10011101, 0b00000001,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 )
        ],
    );

    my $bytes_500 = 'x' x 500;
    push @bytes,
        $bytes_500 => [
        0b10011110, 0b00000000, 0b11010111,
        map { ord($_) } split //, $bytes_500
        ];

    my $bytes_2000 = 'x' x 2000;
    push @bytes,
        $bytes_2000 => [
        0b10011110, 0b00000110, 0b10110011,
        map { ord($_) } split //, $bytes_2000
        ];

    my $bytes_70000 = 'x' x 70000;
    push @bytes,
        $bytes_70000 => [
        0b10011111, 0b00000000, 0b00010000, 0b01010011,
        map { ord($_) } split //, $bytes_70000
        ];

    return \@bytes;
}

sub _double {
    return [
        0.0 => [
            0b01101000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000
        ],
        0.5 => [
            0b01101000,
            0b00111111, 0b11100000, 0b00000000, 0b00000000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000
        ],
        3.14159265359 => [
            0b01101000,
            0b01000000, 0b00001001, 0b00100001, 0b11111011,
            0b01010100, 0b01000100, 0b00101110, 0b11101010
        ],
        123 => [
            0b01101000,
            0b01000000, 0b01011110, 0b11000000, 0b00000000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000
        ],
        1073741824.12457 => [
            0b01101000,
            0b01000001, 0b11010000, 0b00000000, 0b00000000,
            0b00000000, 0b00000111, 0b11111000, 0b11110100
        ],
        -0.5 => [
            0b01101000,
            0b10111111, 0b11100000, 0b00000000, 0b00000000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000
        ],
        -3.14159265359 => [
            0b01101000,
            0b11000000, 0b00001001, 0b00100001, 0b11111011,
            0b01010100, 0b01000100, 0b00101110, 0b11101010
        ],
        -1073741824.12457 => [
            0b01101000,
            0b11000001, 0b11010000, 0b00000000, 0b00000000,
            0b00000000, 0b00000111, 0b11111000, 0b11110100
        ],
    ];
}

sub _float {
    return [
        0.0 => [
            0b00000100, 0b00001000,
            0b00000000, 0b00000000, 0b00000000, 0b00000000
        ],
        1.0 => [
            0b00000100, 0b00001000,
            0b00111111, 0b10000000, 0b00000000, 0b00000000
        ],
        1.1 => [
            0b00000100, 0b00001000,
            0b00111111, 0b10001100, 0b11001100, 0b11001101
        ],
        3.14 => [
            0b00000100, 0b00001000,
            0b01000000, 0b01001000, 0b11110101, 0b11000011
        ],
        9999.99 => [
            0b00000100, 0b00001000,
            0b01000110, 0b00011100, 0b00111111, 0b11110110
        ],
        -1.0 => [
            0b00000100, 0b00001000,
            0b10111111, 0b10000000, 0b00000000, 0b00000000
        ],
        -1.1 => [
            0b00000100, 0b00001000,
            0b10111111, 0b10001100, 0b11001100, 0b11001101
        ],
        -3.14 => [
            0b00000100, 0b00001000,
            0b11000000, 0b01001000, 0b11110101, 0b11000011
        ],
        -9999.99 => [
            0b00000100, 0b00001000,
            0b11000110, 0b00011100, 0b00111111, 0b11110110
        ],
    ];
}

sub _int32 {
    my $max = 2**31 - 1;

    return [
        0 => [ 0b00000000, 0b00000001 ],
        -1 => [
            0b00000100, 0b00000001,
            0b11111111, 0b11111111, 0b11111111, 0b11111111
        ],
        ( 2**8 - 1 ) => [
            0b00000001, 0b00000001,
            0b11111111,
        ],
        -1 * ( 2**8 - 1 ) => [
            0b00000100, 0b00000001,
            0b11111111, 0b11111111, 0b11111111, 0b00000001,
        ],
        500 => [
            0b00000010, 0b00000001,
            0b00000001, 0b11110100
        ],
        -500 => [
            0b00000100, 0b00000001,
            0b11111111, 0b11111111, 0b11111110, 0b00001100,
        ],
        ( 2**16 - 1 ) => [
            0b00000010, 0b00000001,
            0b11111111, 0b11111111
        ],
        -1 * ( 2**16 - 1 ) => [
            0b00000100, 0b00000001,
            0b11111111, 0b11111111, 0b00000000, 0b00000001,
        ],
        ( 2**24 - 1 ) => [
            0b00000011, 0b00000001,
            0b11111111, 0b11111111, 0b11111111
        ],
        -1 * ( 2**24 - 1 ) => [
            0b00000100, 0b00000001,
            0b11111111, 0b00000000, 0b00000000, 0b00000001,
        ],
        $max => [
            0b00000100, 0b00000001,
            0b01111111, 0b11111111, 0b11111111, 0b11111111
        ],
        -1 * $max => [
            0b00000100, 0b00000001,
            0b10000000, 0b00000000, 0b00000000, 0b00000001,
        ]
    ];
}

sub _map {
    my @one_key = (
        { 'en' => 'Foo' } => [
            0b11100001,
            (
                # en
                0b01000010,
                0b01100101, 0b01101110
            ),
            (
                # Foo
                0b01000011,
                0b01000110, 0b01101111, 0b01101111
            )
        ],
    );

    my @two_keys = (
        {
            'en' => 'Foo',
            'zh' => "\x{4eba}",
        } => [
            0b11100010,
            (
                # en
                0b01000010,
                0b01100101, 0b01101110
            ),
            (
                # Foo
                0b01000011,
                0b01000110, 0b01101111, 0b01101111
            ),
            (
                # zh
                0b01000010,
                0b01111010, 0b01101000
            ),
            (
                # \x{4eba}
                0b01000011,
                0b11100100, 0b10111010, 0b10111010
            ),
        ],
    );

    my @nested = (
        { name => { en => 'Bar', zh => "\x{4ebb}" } } => [
            0b11100001,
            (
                # name
                0b01000100,
                0b01101110, 0b01100001, 0b01101101, 0b01100101
            ),
            0b11100010,
            (
                # en
                0b01000010,
                0b01100101, 0b01101110
            ),
            (
                # Bar
                0b01000011,
                0b01000010, 0b01100001, 0b01110010
            ),
            (
                # zh
                0b01000010,
                0b01111010, 0b01101000
            ),
            (
                # \x{4ebb}
                0b01000011,
                0b11100100, 0b10111010, 0b10111011
            ),
        ],
    );

    my @guess = (
        {
            languages => [ 'en', 'zh' ],
        } => [
            0b11100001,
            (    # languages
                0b01001001,
                0b01101100, 0b01100001, 0b01101110, 0b01100111, 0b01110101,
                0b01100001, 0b01100111, 0b01100101, 0b01110011
            ),
            (
                # array
                0b00000010,
                0b00000100,
                (
                    # en
                    0b01000010,
                    0b01100101, 0b01101110
                ),
                (
                    # zh
                    0b01000010,
                    0b01111010, 0b01101000
                ),
            ),
        ],
    );

    return [
        {} => [0b11100000],
        @one_key,
        @two_keys,
        @nested,
        @guess,
    ];
}

sub _pointer {
    return [
        0 => [
            0b00100000,
            0b00000000
        ],
        5 => [
            0b00100000,
            0b00000101
        ],
        10 => [
            0b00100000,
            0b00001010
        ],
        ( 2**10 ) - 1 => [
            0b00100011,
            0b11111111,
        ],
        3017 => [
            0b00101000,
            0b00000011, 0b11001001
        ],
        2**19 - 5 => [
            0b00101111,
            0b11110111, 0b11111011
        ],
        ( 2**19 + 2**11 ) - 1 => [
            0b00101111,
            0b11111111, 0b11111111
        ],
        2**27 - 2 => [
            0b00110111,
            0b11110111, 0b11110111, 0b11111110
        ],
        ( 2**27 + 2**19 + 2**11 ) - 1 => [
            0b00110111,
            0b11111111, 0b11111111, 0b11111111
        ],
        ( 2**32 - 1 ) => [
            0b00111000,
            0b11111111, 0b11111111, 0b11111111, 0b11111111
        ],
    ];
}

sub _uint128 {
    my @uint128 = (
        Math::BigInt->bzero() => [ 0b00000000, 0b00000011 ],
        Math::BigInt->new(500) => [
            0b00000010, 0b00000011,
            0b00000001, 0b11110100
        ],
        Math::BigInt->new(10872) => [
            0b00000010, 0b00000011,
            0b00101010, 0b01111000
        ],
    );

    for my $power ( 1 .. 16 ) {
        my $key = Math::BigInt->new(2)**( 8 * $power ) - Math::BigInt->bone();

        my $value = [
            $power, 0b00000011,
            (0b11111111) x $power
        ];

        push @uint128, $key => $value;
    }

    return \@uint128;
}

sub _uint64 {
    my @uint64 = (
        Math::BigInt->bzero() => [ 0b00000000, 0b00000010 ],
        Math::BigInt->new(500) => [
            0b00000010, 0b00000010,
            0b00000001, 0b11110100
        ],
        Math::BigInt->new(10872) => [
            0b00000010, 0b00000010,
            0b00101010, 0b01111000
        ],
    );

    for my $power ( 1 .. 8 ) {
        my $key = Math::BigInt->new(2)**( 8 * $power ) - Math::BigInt->bone();

        my $value = [
            $power, 0b00000010,
            (0b11111111) x $power
        ];

        push @uint64, $key => $value;
    }

    return \@uint64;
}

sub _uint32 {
    my $max = 2**32 - 1;

    return [
        0            => [0b11000000],
        ( 2**8 - 1 ) => [
            0b11000001,
            0b11111111,
        ],
        500 => [
            0b11000010,
            0b00000001, 0b11110100
        ],
        10872 => [
            0b11000010,
            0b00101010, 0b01111000
        ],
        ( 2**16 - 1 ) => [
            0b11000010,
            0b11111111, 0b11111111
        ],
        ( 2**24 - 1 ) => [
            0b11000011,
            0b11111111, 0b11111111, 0b11111111
        ],
        $max => [
            0b11000100,
            0b11111111, 0b11111111, 0b11111111, 0b11111111
        ],
    ];
}

sub _uint16 {
    my $max = 2**16 - 1;

    return [
        0            => [0b10100000],
        ( 2**8 - 1 ) => [
            0b10100001,
            0b11111111,
        ],
        500 => [
            0b10100010,
            0b00000001, 0b11110100
        ],
        10872 => [
            0b10100010,
            0b00101010, 0b01111000
        ],
        $max => [
            0b10100010,
            0b11111111, 0b11111111,
        ],
    ];
}

sub _utf8_string {
    shift;
    my %p = @_;

    my @string = (
        q{}        => [0b01000000],
        '1'        => [ 0b01000001, ord('1') ],
        "\x{4eba}" => [
            0b01000011,
            0b11100100, 0b10111010, 0b10111010
        ],
        '123' => [ 0b01000011, map { ord($_) } qw( 1 2 3 ) ],
        '123456789012345678901234567' => [
            0b01011011,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 )
        ],
        '1234567890123456789012345678' => [
            0b01011100,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 )
        ],
        '12345678901234567890123456789' => [
            0b01011101, 0b00000000,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 )
        ],
        '123456789012345678901234567890' => [
            0b01011101, 0b00000001,
            map { ord($_) }
                qw( 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 )
        ],
    );

    my $string_500 = 'x' x 500;
    push @string,
        $string_500 => [
        0b01011110, 0b00000000, 0b11010111,
        map { ord($_) } split //, $string_500
        ];

    my $string_2000 = 'x' x 2000;
    push @string,
        $string_2000 => [
        0b01011110, 0b00000110, 0b10110011,
        map { ord($_) } split //, $string_2000
        ];

    my $string_70000 = 'x' x 70000;
    push @string,
        $string_70000 => [
        0b01011111, 0b00000000, 0b00010000, 0b01010011,
        map { ord($_) } split //, $string_70000
        ];

    return \@string;
}

sub _boolean {
    return [
        0 => [
            0b00000000, 0b00000111,
        ],
        1 => [
            0b00000001, 0b00000111,
        ],
    ];
}
## use critic

1;


lib/Test/MaxMind/DB/Common/Util.pm
package Test::MaxMind::DB::Common::Util;

use strict;
use warnings;

our $VERSION = '0.040002';

use Exporter qw( import );

our @EXPORT_OK = qw( standard_test_metadata );

sub standard_test_metadata {
    return (
        database_type => 'Test',
        languages     => [ 'en', 'zh' ],
        description   => {
            en => 'Test Database',
            zh => 'Test Database Chinese',
        },
    );
}

1;


lib/MaxMind/DB/Types.pm
package MaxMind::DB::Types;

use strict;
use warnings;

our $VERSION = '0.040002';

use Carp qw( confess );
use Exporter qw( import );
use List::AllUtils;
use Scalar::Util ();
use Sub::Quote qw( quote_sub );
use overload ();

our @EXPORT_OK = qw(
    ArrayRefOfStr
    Bool
    Decoder
    Epoch
    FileHandle
    HashRef
    HashRefOfStr
    Int
    Metadata
    Str
);

## no critic (NamingConventions::Capitalization, ValuesAndExpressions::ProhibitImplicitNewlines)
{
    my $t = quote_sub(
        q{
(
           defined $_[0]
        && Scalar::Util::reftype( $_[0] ) eq 'ARRAY'
        && List::AllUtils::all(
        sub { defined $_ && !ref $_ },
        @{ $_[0] }
        )
    )
    or MaxMind::DB::Types::_confess(
    '%s is not an arrayref',
    $_[0]
    );
}
    );

    sub ArrayRefOfStr () { $t }
}

{
    my $t = quote_sub(
        q{
( !defined $_[0] || $_[0] eq q{} || "$_[0]" eq '1' || "$_[0]" eq '0' )
    or MaxMind::DB::Types::_confess(
    '%s is not a boolean',
    $_[0]
    );
}
    );

    sub Bool () { $t }
}

{
    my $t = _object_isa_type('MaxMind::DB::Reader::Decoder');

    sub Decoder () { $t }
}

{
    my $t = quote_sub(
        q{
(
    defined $_[0] && ( ( !ref $_[0] && $_[0] =~ /^[0-9]+$/ )
        || ( Scalar::Util::blessed( $_[0] )
            && ( $_[0]->isa('Math::UInt128') || $_[0]->isa('Math::BigInt') ) )
        )
    )
    or MaxMind::DB::Types::_confess(
    '%s is not an integer, a Math::UInt128 object, or a Math::BigInt object',
    $_[0]
    );
}
    );

    sub Epoch () { $t }

    sub Int () { $t }
}

{
    my $t = quote_sub(
        q{
(          ( defined $_[0] && Scalar::Util::openhandle( $_[0] ) )
        || ( Scalar::Util::blessed( $_[0] ) && $_[0]->isa('IO::Handle') ) )
    or MaxMind::DB::Types::_confess(
    '%s is not a file handle',
    $_[0]
    );
}
    );

    sub FileHandle () { $t }
}

{
    my $t = quote_sub(
        q{
( defined $_[0] && Scalar::Util::reftype( $_[0] ) eq 'HASH' )
    or MaxMind::DB::Types::_confess(
    '%s is not a hashref',
    $_[0]
    );
}
    );

    sub HashRef () { $t }
}

{
    my $t = quote_sub(
        q{
(
           defined $_[0]
        && Scalar::Util::reftype( $_[0] ) eq 'HASH'
        && &List::AllUtils::all(
        sub { defined $_ && !ref $_ }, values %{ $_[0] }
        )
    )
    or MaxMind::DB::Types::_confess(
    '%s is not a hashref of strings',
    $_[0]
    );
}
    );

    sub HashRefOfStr () { $t }
}

{
    my $t = _object_isa_type('MaxMind::DB::Metadata');

    sub Metadata () { $t }
}

{
    my $t = quote_sub(
        q{
( defined $_[0] && !ref $_[0] )
    or MaxMind::DB::Types::_confess( '%s is not a string', $_[0] );
}
    );

    sub Str () { $t }
}

sub _object_isa_type {
    my $class = shift;

    return quote_sub(
        qq{
( Scalar::Util::blessed( \$_[0] ) && \$_[0]->isa('$class') )
    or MaxMind::DB::Types::_confess(
    '%s is not a $class object',
    \$_[0]
    );
}
    );
}
## use critic

## no critic (Subroutines::ProhibitUnusedPrivateSubroutines)
sub _confess {
    ## no critic (Subroutines::ProhibitCallsToUnexportedSubs)
    confess sprintf(
        $_[0],
        defined $_[1] ? overload::StrVal( $_[1] ) : 'undef'
    );
}
## use critic

1;


lib/MaxMind/DB/Metadata.pm
package MaxMind::DB::Metadata;

use strict;
use warnings;
use namespace::autoclean;

our $VERSION = '0.040002';

use Moo;
use MaxMind::DB::Types qw( ArrayRefOfStr Epoch HashRefOfStr Int Str );
use MooX::StrictConstructor;

with 'MaxMind::DB::Role::Debugs';

{
    my %metadata = (
        binary_format_major_version => Int,
        binary_format_minor_version => Int,
        build_epoch                 => Epoch,
        database_type               => Str,
        description                 => HashRefOfStr,
        ip_version                  => Int,
        node_count                  => Int,
        record_size                 => Int,
    );

    for my $attr ( keys %metadata ) {
        has $attr => (
            is       => 'ro',
            isa      => $metadata{$attr},
            required => 1,
        );
    }
}

has languages => (
    is      => 'ro',
    isa     => ArrayRefOfStr,
    default => sub { [] },
);

sub metadata_to_encode {
    my $self = shift;

    my %metadata;
    foreach my $attr ( $self->meta()->get_all_attributes() ) {
        my $method = $attr->name;
        $metadata{$method} = $self->$method;
    }

    return \%metadata;
}

sub debug_dump {
    my $self = shift;

    $self->_debug_newline();

    $self->_debug_message('Metadata:');
    my $version = join '.',
        $self->binary_format_major_version(),
        $self->binary_format_minor_version();
    $self->_debug_string( '  Binary format version', $version );

    require DateTime;
    $self->_debug_string(
        '  Build epoch',
        $self->build_epoch() . ' ('
            . DateTime->from_epoch( epoch => $self->build_epoch() ) . ')'
    );

    $self->_debug_string( '  Database type', $self->database_type() );

    my $description = $self->description();
    for my $locale ( sort keys %{$description} ) {
        $self->_debug_string(
            "  Description [$locale]",
            $description->{$locale}
        );
    }

    $self->_debug_string( '  IP version',            $self->ip_version() );
    $self->_debug_string( '  Node count',            $self->node_count() );
    $self->_debug_string( '  Record size (in bits)', $self->record_size() );
    $self->_debug_string(
        '  Languages', join ', ',
        @{ $self->languages() }
    );

    return;
}

__PACKAGE__->meta()->make_immutable();

1;

#ABSTRACT: A class for metadata related to a MaxMind DB database

__END__

=for test_synopsis
my $path;

=head1 SYNOPSIS

    my $reader = MaxMind::DB::Reader->new( file => $path );
    my $metadata = $reader->metadata();

    print $metadata->description()->{en};

=head1 DESCRIPTION

This class provides an API for representing the metadata of a MaxMind DB
database. See http://maxmind.github.io/MaxMind-DB/ for the official format
spec.

=head1 API

This class provides methods for each metadata attribute in a database.

=head2 $metadata->binary_format_major_version()

Returns the binary format major version number.

=head2 $metadata->binary_format_minor_version()

Returns the binary format minor version number.

=head2 $metadata->build_epoch()

Returns the database's build timestamp as an epoch value.

=head2 $metadata->database_type()

Returns a string indicating the database's type.

=head2 $metadata->languages()

Returns an arrayref of locale codes indicating what languages this database
has information for.

=head2 $metadata->description()

Returns a hashref of descriptions. The keys should be locale codes like "en"
or "pt-BR" and the values are the description in that language.

=head2 $metadata->ip_version()

Returns a 4 or 6 indicating what type of IP addresses this database can be
used to look up.

=head2 $metadata->node_count()

Returns the number of nodes in the database's search tree.

=head2 $metadata->record_size()

Returns the record size for nodes in the database's search tree.


lib/MaxMind/DB/Role/Debugs.pm
package MaxMind::DB::Role::Debugs;

use strict;
use warnings;
use namespace::autoclean;
use autodie;

our $VERSION = '0.040002';

use Moo::Role;

sub _debug_newline {
    warn "\n";
}

sub _debug_binary {
    my $self   = shift;
    my $label  = shift;
    my $binary = shift;

    my $msg = "$label: ";

    if ( defined $binary ) {
        $msg .= join q{ }, map { sprintf( '%08b', ord($_) ) } split //,
            $binary;
    }
    else {
        $msg .= '<undef>';
    }

    warn "$msg\n";
}

sub _debug_string {
    my $self   = shift;
    my $label  = shift;
    my $string = shift;

    $string = '<undef>'
        unless defined $string;

    warn "$label: $string\n";
}

sub _debug_sprintf {
    my $self = shift;

    warn sprintf( shift() . "\n", @_ );
}

sub _debug_structure {
    my $self      = shift;
    my $label     = shift;
    my $structure = shift;

    require Data::Dumper::Concise;
    warn "$label: \n";
    my $dumped = Data::Dumper::Concise::Dumper($structure);

    $dumped =~ s/^/  /mg;

    warn $dumped;
}

sub _debug_message {
    my $self = shift;
    my $msg  = shift;

    warn "$msg\n";
}

1;


lib/MaxMind/DB/Common.pm
package MaxMind::DB::Common;

use strict;
use warnings;

our $VERSION = '0.040002';

my $separator_size;

BEGIN {
    $separator_size = 16;
}

use constant {
    LEFT_RECORD                 => 0,
    RIGHT_RECORD                => 1,
    DATA_SECTION_SEPARATOR_SIZE => $separator_size,
    DATA_SECTION_SEPARATOR      => ( "\0" x $separator_size ),
    METADATA_MARKER             => "\xab\xcd\xefMaxMind.com",
};

use Exporter qw( import );

## no critic (Variables::ProhibitPackageVars)
our %TypeNumToName = (
    0  => 'extended',
    1  => 'pointer',
    2  => 'utf8_string',
    3  => 'double',
    4  => 'bytes',
    5  => 'uint16',
    6  => 'uint32',
    7  => 'map',
    8  => 'int32',
    9  => 'uint64',
    10 => 'uint128',
    11 => 'array',
    12 => 'container',
    13 => 'end_marker',
    14 => 'boolean',
    15 => 'float',
);

our %TypeNameToNum = reverse %TypeNumToName;

our @EXPORT_OK = qw(
    LEFT_RECORD
    RIGHT_RECORD
    DATA_SECTION_SEPARATOR_SIZE
    DATA_SECTION_SEPARATOR
    METADATA_MARKER
    %TypeNumToName
    %TypeNameToNum
);

1;

# ABSTRACT: Code shared by the MaxMind DB reader and writer modules

__END__

=head1 DESCRIPTION

This distribution provides some shared code for use by both the MaxMind DB
reader and writer Perl modules.

For now, the only piece documented for public consumption is
L<MaxMind::DB::Metadata>.

=head1 VERSIONING POLICY

This module uses semantic versioning as described by
L<http://semver.org/>. Version numbers can be read as X.YYYZZZ, where X is the
major number, YYY is the minor number, and ZZZ is the patch number.

=head1 SUPPORT

Please report all issues with this code using the GitHub issue tracker at
L<https://github.com/maxmind/MaxMind-DB-Common-perl/issues>.
